### PR TITLE
Set our own game language

### LINKED ISSUE #124: Language
How to change language?

### PR DESCRIPTION
**Pull request by jokeruarwentto**
Allow the players to set their own language with an ``language.txt`` file in game's root (where ``h1-mod.exe`` is).

image

(edit: closes #124)

### CODE REVIEW

File: src/client/steam/interfaces/apps.hpp [hunk 1]
```diff
@@ -7,6 +7,29 @@ namespace steam
...
+				while(std::getline(file, line))
+					Language = line.c_str();
+
+#if DEBUG
+				printf("[Language] Switched to \"%s\".\n", Language);
+#endif
+			}
+		}
+
+		const char* LanguageFileName = "language.txt";
```
Comment (fedddddd) [inline src/client/steam/interfaces/apps.hpp:30]: No need for this to be a member

File: src/client/steam/interfaces/apps.hpp [hunk 2]
```diff
@@ -7,6 +7,29 @@ namespace steam
...
+		const char* Language = "english";
```
Comment (fedddddd) [inline src/client/steam/interfaces/apps.hpp:31]: Snake case

File: src/client/steam/interfaces/apps.hpp [hunk 3]
```diff
@@ -7,6 +7,29 @@ namespace steam
...
+			{
+#if DEBUG
+				printf("[Language] Custom language file has been found.\n");
+#endif
+
+				std::ifstream file;
+				file.open(LanguageFileName);
+				std::string line;
+				while(std::getline(file, line))
+					Language = line.c_str();
```
Comment (fedddddd) [inline src/client/steam/interfaces/apps.hpp:22]: Just use `utils::io::file_exists` and `utils::io::read_file`.
Also `std::string line` is a local variable which will be deleted after the execution of `apps()` so you should make `const char* Language` a char array and copy the result of `utils::io::read_file` to it

**Review by fedddddd — CHANGES_REQUESTED** [review]

**Comment by VladWinner** [comment]
AFAIK, the game picks up another language if you delete the zone/english folder or rename it to english_
This does not affect the client's functioning.

**Comment by mjkzy** [comment]
> [Workaround] AFAIK, the game just picks up another language if you delete the `zone/english` folder or rename it to `english_`
> 
> This does not affect the client's functioning.

even if that's the case, having this language choosing is pretty good imo. i'll fix this PR up later and merge

**Review by fedddddd — CHANGES_REQUESTED** [review]

File: data/ui_scripts/language_setting/__init__.lua [hunk 4]
```diff
@@ -0,0 +1,153 @@
+local localization = {
+    english = {
+        LANGUAGE_BUTTON = "Language",
+        LANGUAGE_BUTTON_DESC = "Change your game language.",
+        POPUP_RESTART_REQUIRED_TITLE = "RESTART REQUIRED",
+        POPUP_RESTART_REQUIRED_TEXT = "WARNING: To work properly, you need to have your language into zone directory of your game root. You have changed your game language, wish you restart?",
```
Comment (mjkzy) [inline data/ui_scripts/language_setting/__init__.lua:6]: "... you need to have your language's zone files in the zone directory ..."

"... do you wish to restart? ..."

File: data/ui_scripts/language_setting/__init__.lua [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        POPUP_NO_ZONE_FOUND_TITLE = "LANGUAGE NOT FOUND",
+        POPUP_NO_ZONE_FOUND_TEXT = "We are unable to find the language folder you have selected, to avoid bugs, we will not change anything.",
```
Comment (mjkzy) [inline data/ui_scripts/language_setting/__init__.lua:8]: "... you have selected. To avoid bugs, ..."

**Review by mjkzy — COMMENTED** [review]
you actually caught us right before we merged this PR, haha. we were gonna ship this with the next release but we postponed it to the next one after.

i like the usage of adding LUI into the question and i think we can go ahead and add this but i think the bigger issue is the fact that people's game will crash before even trying to get into game to change the option. would you agree?

File: data/ui_scripts/language_setting/__init__.lua [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+function set_language(value)
+    local file_path = "h1-mod/language.txt"
+    local file = io.open(file_path, "w")
+    file:write(value)
+    file:close()
+end
+
+function does_zone_folder_exists(language)
+    local isok, errstr, errcode = os.rename("zone/" .. language .. "/", "zone/" .. language .. "/")
```
Comment (mjkzy) [inline data/ui_scripts/language_setting/__init__.lua:90]: forgot to add this to the review - does renaming show if the folder actually exists or something?

if it does, consider using io::directoryexists

Comment (mjkzy) [reply data/ui_scripts/language_setting/__init__.lua:90]: we add it on the client [here](https://github.com/h1-mod/h1-mod/blob/develop/src/client/component/ui_scripting.cpp#L176) :)

**Comment by mjkzy** [comment]
the images shown above look good, i must have been thinking of another menu

**Comment by mjkzy** [comment]
> Yes, we may need to check if english zone folder exists else it automatically take the first language they've ? But 90% peoples, I guess, got the game through your services

it's not even that. if someone doesn't have english files but they have spanish, for example, and they try to launch, it'll default to english. however, they don't have english. this is where a random error occurs with missing content.
my solution? i'd say modify the 'MISSING CONTENT' error we have in the client and do a search of the zone directory to see if any known languages exist. if they do, then create that language.txt file and tell them to restart your game with something like "We've detected another language installed. Please restart your game to fix this error."

**Comment by mjkzy** [comment]
if you wanna go ahead and try to do this, we can expand on it after revision! :)

**Comment by mjkzy** [comment]
oh wow, didn't know that. does the apps() handler shit still default to "english" though?

**Comment by VladWinner** [comment]
`local available_languages = {"english", "french", "italian", "german", "spanish"}`
@jokeruarwentto Can you add `russian` localization to the list?

**Comment by VladWinner** [comment]
> I guess I will need help to translate it aha
[__init__ (ru).txt](https://github.com/h1-mod/h1-mod/files/8996108/__init__.ru.txt)

> English font does not like it at all!
The original font does not contain Cyrillic (33 uppercase + 33 lowercase letters), so in the Russian localization game fonts are globally replaced by one Arial-like font >.<

**Comment by VladWinner** [comment]
> Yep, we need to do something there... 😞

I thought it would be more logical if this menu is displayed in the current language. So there will be no conflicts with partial character coverage in the original font.

**Comment by mjkzy** [comment]
what's your discords? I can add you guys to a channel to discuss the future of this PR in the h1-mod discord

**Comment by VladWinner** [comment]
> Arwent#0001

ruVlad#7778

**Comment by mjkzy** [comment]
are we good to merge this PR?

**Pull request merged**